### PR TITLE
[Core] Add MapReducer to core

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This adds `MapReducer` to core at `ReductionUtilities`.

**🆕 Changelog**
- Adds `MapReducer`
- Adds test

### CODE REVIEW

**Review by philbucher — COMMENTED** [review]
I wrote this already somewhere, please wait with merging until I found it

File: kratos/utilities/reduction_utilities.h [hunk 1]
```diff
@@ -278,6 +278,36 @@ class AccumReduction
     }
 };
 
+template<class MapType>
+class MapReduction
```
Comment (philbucher) [inline kratos/utilities/reduction_utilities.h:282]: Please check with what I added in #10339, I would think that `merge` is faster

In any case, can you then please also update the custom solution?

Comment (sunethwarna) [reply kratos/utilities/reduction_utilities.h:282]: done :D

**Review by philbucher — DISMISSED** [review]
Thx

**Review by philbucher — APPROVED** [review]

**Pull request merged**